tests/helpers/test_entity_registry.py

```diff
@@ -222,3 +223,36 @@ def test_async_get_entity_id(registry):
     assert entry.name == 'Test Name'
     assert entry.disabled_by == 'hass'
     assert entry.config_entry_id == 'test-config-id'
+
+
+async def test_loading_invalid_entity_id(hass, hass_storage):
+    """Test we autofix invalid entity IDs."""
+    hass_storage[entity_registry.STORAGE_KEY] = {
+        'version': entity_registry.STORAGE_VERSION,
+        'data': {
+            'entities': [
+                {
+                    'entity_id': 'test.invalid__middle',
+                    'platform': 'super_platform',
+                    'unique_id': 'id-invalid-middle',
+                    'name': 'registry override',
+                }, {
+                    'entity_id': 'test.invalid_end_',
+                    'platform': 'super_platform',
+                    'unique_id': 'id-invalid-end',
+                }
+            ]
+        }
+    }
+
+    registry = await entity_registry.async_get_registry(hass)
+
+    entity_invalid_middle = registry.async_get_or_create(
+        'test', 'super_platform', 'id-invalid-middle')
+
+    assert valid_entity_id(entity_invalid_middle.entity_id)
+
+    entity_invalid_end = registry.async_get_or_create(
```

Comment:
`slugify('_nostart') = 'nostart'` but is covered by your implementation of new_entity_id

Reply:
https://github.com/home-assistant/home-assistant/pull/20357